### PR TITLE
Update Maps SDK with MAS 3.0

### LINKED ISSUE #746: Usage of mapboxServices vs mapboxSdkServices
We are including two different versions of [mapbox-java](https://github.com/mapbox/mapbox-java/) in this project. See [here](https://github.com/mapbox/mapbox-navigation-android/blob/master/gradle/dependencies.gradle#L12-L13) and [here](https://github.com/mapbox/mapbox-navigation-android/blob/master/libandroid-navigation/build.gradle#L46-L48). Considering that we are using the `Maps SDK v5.5.0` we should remove `mapbox-java v3.0.0-beta.3`, that version of is only compatible with the beta versions of the 6.0.0 release or update maps to 6.0.0-beta.x and remove MAS 2.10.0.

### LINKED ISSUE #837: Passing a null gesture event listener
Refs. https://github.com/mapbox/mapbox-gestures-android/issues/38 and https://github.com/mapbox/mapbox-gl-native/issues/11628.

Clearing gestures events listeners list is not support with Maps SDK `6.0.0`. Operations like https://github.com/mapbox/mapbox-navigation-android/blob/5a5ae26c22e5b02b68d43e8d8aa08c316293626a/app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/MockNavigationActivity.java#L129
are not allowed.

### LINKED ISSUE #880: Mapbox zoom LatLngBound sometimes wrong
When I use  navigationView for my project, i have a function to zoom map in the center of start point and end point:

navigationView..getMapboxMap().easeCamera(CameraUpdateFactory.newLatLngBounds(latLngBounds, 10));

but the strange thing happen: some time it zoom right, some time it zoom wrong (not move to center of LatLngBound).

### LINKED ISSUE #891: NavigationView Crashes issue, binary xml file #0.
I am writing this issue because I believe someone in the future may have the same issue and I fixed it finally with some efforts.

The error is NavigationView from Navigation UI crashes with Binary XML file error. Here is what I have done and the short reason is that **android telemetry version 3.0.2 and 2.2.10 API are inconsistent**.

**Step1: use plugin for searching place with UI**

The plugin-places provides you an easy way to handle user's search for places with a decent autocomplete textview UI.
https://github.com/mapbox/mapbox-plugins-android/tree/master/plugin-places

from there, it tells you to import:
    implementation 'com.mapbox.mapboxsdk:mapbox-android-plugin-places:0.3.0'

As a result, one following library called telemetry-3.0.2 is imported automatically. 
image

Please notice here the version is 3.0.2 here.

**Step2: use navigation sdk and UI**
Now you want to do navigation task after user's searching.
So you import android navigation SDK:
https://github.com/mapbox/mapbox-navigation-android

compile 'com.mapbox.mapboxsdk:mapbox-android-navigation:0.12.0'

along with the UI library:

compile ('com.mapbox.mapboxsdk:mapbox-android-navigation-ui:0.12.0')

**Step3: Error happens when you launch the navigationview**
When you launch the navigationView, you will have the error something like that: 

Binary XML file line #0
Error inflating class com.mapbox.services.android.navigation.ui.v5.NavigationView
  Caused by: android.view.InflateException: Binary XML file line #0: Error inflating class com.mapbox.services.android.navigation.ui.v5.NavigationView
...
And the last row tells you that:
**com.mapbox.services.android.telemetry.location does not exist**

Because navigationview also uses telemetry library for localization, it will not arise such issue in compile time while in run time it will have a class not found error.

**Reasons:**
If you only import 
com.mapbox.mapboxsdk:mapbox-android-navigation-ui:0.12.0
com.mapbox.mapboxsdk:mapbox-android-navigation:0.12.0

It will automatically import required telemetry library:
image
And in this version, it has com.mapbox.services.android.telemetry.location class. While in version  
3.0.2, it changes the API and no longer has such class. (Far more than this)

**Solution and Questions**
This is actually the quesion I want to ask, If there is some elegent way to keep the library versions consistent?

Also a suggestion could be that could you add some alerts when using multiple library in the same time, some versions may conflict, especially for those that silently imported.

### PR DESCRIPTION
**Pull request by danesfeder**
Closes #891 Closes #880 Closes #837 Closes #746 

Currently not compiling due to a dex error with something related to offline

cc @osana @cammace (for 👀 on the `NavigationMapRoute` code)

### CODE REVIEW

**Comment by danesfeder** [comment]
@tobrun any ideas here?  

```
Error:com.android.builder.dexing.DexArchiveBuilderException: Failed to process /Users/danesfeder/.gradle/caches/transforms-1/files-1.1/mapbox-android-sdk-6.0.0-SNAPSHOT.aar/0db8873ff2b10dc56820a1208701e197/jars/classes.jar

Error:com.android.builder.dexing.DexArchiveBuilderException: Error while dexing com/mapbox/mapboxsdk/offline/OfflineRegion$2.class

Error:com.android.dx.cf.code.SimException: invalid opcode ba (invokedynamic requires --min-sdk-version >= 26)```

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java [hunk 1]
```diff
@@ -553,20 +549,19 @@ private void drawWaypointMarkers(@NonNull MapboxMap mapboxMap, @Nullable Drawabl
...
       waypointLayer = new SymbolLayer(WAYPOINT_LAYER_ID, WAYPOINT_SOURCE_ID).withProperties(
-        PropertyFactory.iconImage(Function.property(
-          "waypoint",
-          categorical(
-            stop("origin", PropertyFactory.iconImage("originMarker")),
-            stop("destination", PropertyFactory.iconImage("destinationMarker"))
+        PropertyFactory.iconImage(match(
+          get("waypoint"),
+          literal("originMarker"), literal("stop"), literal("originMarker"),
+          literal("destination"), literal("destinationMarker")
```
Comment (tobrun) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java]: This conversion doesn't seem correct:

match(
 get("waypoint"),
 literal("origin"),  literal("originMarker"),
 literal("destination"), literal("destinationMarker"),
 literal("defaultOutputIfAboveDoesntMatch")
)

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/MainActivity.java [hunk 2]
```diff
@@ -13,20 +13,21 @@
...
-import com.mapbox.services.android.telemetry.permissions.PermissionsManager;
 
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 
 public class MainActivity extends AppCompatActivity implements PermissionsListener {
 
+  // TODO Check and remove if not necessary
   private static final String[] PERMISSIONS = {Manifest.permission.ACCESS_FINE_LOCATION,
```
Comment (danesfeder) [inline app/src/main/java/com/mapbox/services/android/navigation/testapp/MainActivity.java]: Confirmed: can be removed 👍

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/MockNavigationActivity.java [hunk 3]
```diff
@@ -142,8 +143,9 @@ private void newOrigin() {
         Point.fromLngLat(latLng.getLongitude(), latLng.getLatitude())
       );
       mapboxMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 12));
-      mapboxMap.setMyLocationEnabled(true);
-      mapboxMap.getTrackingSettings().setMyLocationTrackingMode(MyLocationTracking.TRACKING_FOLLOW);
+      // TODO Check and remove if not necessary
```
Comment (danesfeder) [inline app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/MockNavigationActivity.java]: Confirmed deprecated code: can be removed 👍

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/MockNavigationActivity.java [hunk 4]
```diff
@@ -181,7 +184,8 @@ public void onMapClick(@NonNull LatLng point) {
   }
 
   private void calculateRoute() {
-    Location userLocation = mapboxMap.getMyLocation();
+    // TODO Check
+    Location userLocation = locationEngine.getLastLocation();
```
Comment (danesfeder) [inline app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/MockNavigationActivity.java]: Confirmed this is good to replace 👍

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationEventDispatcher.java [hunk 5]
```diff
@@ -32,7 +33,7 @@
   NavigationEventDispatcher() {
     navigationEventListeners = new ArrayList<>();
     milestoneEventListeners = new ArrayList<>();
-    progressChangeListeners = new ArrayList<>();
+    progressChangeListeners = new CopyOnWriteArrayList<>();
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationEventDispatcher.java]: Let's do the same for the rest of the lists, great catch here 💯

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MathUtils.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
+package com.mapbox.services.android.navigation.v5.utils;
+
+
+// TODO Check and remove if not necessary
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MathUtils.java]: If we are using these methods in the nav SDK and they are no longer provided by the libraries we are using, let's add them to our `utils`

**Comment by danesfeder** [comment]
@Guardiola31337 taking a look at the map route changes and found that the start/end points are missing now in this branch: 

screenshot_1520979905

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java [hunk 7]
```diff
@@ -557,20 +549,20 @@ private void drawWaypointMarkers(@NonNull MapboxMap mapboxMap, @Nullable Drawabl
...
 
       waypointLayer = new SymbolLayer(WAYPOINT_LAYER_ID, WAYPOINT_SOURCE_ID).withProperties(
-        PropertyFactory.iconImage(Function.property(
-          "waypoint",
-          categorical(
-            stop("origin", PropertyFactory.iconImage("originMarker")),
-            stop("destination", PropertyFactory.iconImage("destinationMarker"))
+        PropertyFactory.iconImage(match(
+          get("waypoint"),
+          literal("originMarker"), literal("originMarker"),
```
Comment (tobrun) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java]: the first originMarker should be "origin"

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java [hunk 8]
```diff
@@ -557,20 +549,20 @@ private void drawWaypointMarkers(@NonNull MapboxMap mapboxMap, @Nullable Drawabl
...
+          literal("destination"), literal("destinationMarker"),
+          literal("defaultOutputIfAboveDoesntMatch")
```
Comment (tobrun) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java]: it's better to actually assign a real icon id value for above, eg. originMarker

**Comment by Guardiola31337** [comment]
While testing the integration noticed that routes are not drawn properly 👇

draw_routes_issue

This is the `MockNavigationActivity` from the test app (tapping the map twice). Note that black color in the route after placing the second waypoint.

Comparing with `master` (Maps SDK `5.5.0` at the moment of writing) 👇

draw_routes_5_5

it seems that we're messing up with _Expressions_ related code. Going to dig deeper. BTW, it'd be great @tobrun if you could take a look at it as well and let us know if you 👀 something missing or wrong. Thanks 🙏

**Comment by tobrun** [comment]
> it seems that we're messing up with Expressions related code. Going to dig deeper. BTW, it'd be great @tobrun if you could take a look at it as well and let us know if you 👀 something missing or wrong. Thanks 

A quick glance over the code found [here](https://github.com/mapbox/mapbox-navigation-android/pull/697/files#diff-cce5da198b8d3c30274778f11300d70b) didn't flag anything to me. 
Could you specify/link which layer configuration is not working correctly?

**Comment by Guardiola31337** [comment]
Noting here that when testing https://github.com/mapbox/mapbox-plugins-android/issues/424#issuecomment-381241909 noticed some building conflicts 👀
```
* What went wrong:
Execution failed for task ':app:preDebugBuild'.
> Android dependency 'android.arch.lifecycle:runtime' has different version for the compile (1.0.3) and runtime (1.1.1) classpath. You should manually set the same version via DependencyResolution
```

We aren't using `1.1.1` (LLP included it in https://github.com/mapbox/mapbox-plugins-android/pull/419) so in order to avoid them we needed to include `implementation "android.arch.lifecycle:extensions:1.1.1"` in the app's `build.gradle` file. We should fix it in the Nav side though (bumping the version). Ticket to follow.

**Comment by tobrun** [comment]
Note that with final release of LLP this shouldn't be an issue anymore, with https://github.com/mapbox/mapbox-plugins-android/pull/419 we limited the scope of the included dependency.

**Comment by Guardiola31337** [comment]
@tobrun 
> Note that with final release of LLP this shouldn't be an issue anymore, with mapbox/mapbox-plugins-android#419 we limited the scope of the included dependency.

What do you mean with `we limited the scope of the included dependency`?

I think this will be an issue in the Nav side (when including LLP final release dependency) until #864 gets addressed.

**Comment by danesfeder** [comment]
@Guardiola31337 I'd look at `NavigationMapRoute#generateFeatureCollectionList`.  I'm wondering if the issue with the route line has to do with how we generate the features a second time, or with waypoints.

**Comment by Guardiola31337** [comment]
old_current_events_libs_comparison_whole_route

<span style="color:green">Green</span> 👉 New implementation
<span style="color:yellow">Yellow</span> 👉 Old implementation

In terms of numbers we're getting 99 events with the old implementation whereas 95 events with the new implementation. Above the green points can't be seen because they're right under the yellow ones (i.e. same exact location). As you can see below the differences only happened at the beginning right before starting the navigation session.

old_current_events_libs_comparison_beginning

Here we're only 🕵️ the location events, it'd be great to do the same with the Nav-related events.

BTW, based on what we're 👀 with the location events seems legit to go ahead and merge the PR. We can revisit/continue analyzing the navigation events later.

cc @danesfeder @zugaldia

**Pull request merged**